File: data-skill/templates/r/clean_data.R
#!/usr/bin/env Rscript
# Data Cleaning Template
# Purpose: Handle missing values, duplicates, and outliers

library(tidyverse)

# ============================================================================
# USER INPUT: Modify these variables
# ============================================================================
data_file <- "data.csv"
# ============================================================================

# Load data
df <- read_csv(data_file)

cat("\n========================================\n")
cat("DATA CLEANING REPORT\n")
cat("========================================\n\n")

# Initial dimensions
cat("Original dimensions: ", nrow(df), " rows x ", ncol(df), " cols\n\n")

# ============================================================================
# 1. Remove Duplicates
# ============================================================================
df_clean <- df %>% distinct()
removed_dup <- nrow(df) - nrow(df_clean)
cat("Duplicates removed: ", removed_dup, "\n")

# ============================================================================
# 2. Handle Missing Values
# ============================================================================
missing_by_col <- df_clean %>%
  summarise(across(everything(), ~ sum(is.na(.)))) %>%
  pivot_longer(cols = everything(), names_to = "variable", values_to = "missing_count") %>%
  filter(missing_count > 0) %>%
  arrange(desc(missing_count))

if (nrow(missing_by_col) > 0) {
  cat("\nMissing values by column:\n")
  print(missing_by_col)
} else {
  cat("\nNo missing values found.\n")
}

# Option: Drop rows with missing values
# df_clean <- df_clean %>% drop_na()

# Option: Fill missing values
# df_clean <- df_clean %>% mutate(across(where(is.numeric), ~ifelse(is.na(.), mean(., na.rm = TRUE), .)))

# ============================================================================
# 3. Detect Outliers (for numeric columns)
# ============================================================================
numeric_cols <- names(df_clean)[sapply(df_clean, is.numeric)]

if (length(numeric_cols) > 0) {
  cat("\nOutlier detection (IQR method):\n")
  for (col in numeric_cols) {
    Q1 <- quantile(df_clean[[col]], 0.25, na.rm = TRUE)
    Q3 <- quantile(df_clean[[col]], 0.75, na.rm = TRUE)
    IQR <- Q3 - Q1
    lower <- Q1 - 1.5 * IQR
    upper <- Q3 + 1.5 * IQR
    outliers <- sum(df_clean[[col]] < lower | df_clean[[col]] > upper, na.rm = TRUE)
    if (outliers > 0) {
      cat("  ", col, ": ", outliers, " outliers\n")
    }
  }
}

# Final dimensions
cat("\nClean dimensions: ", nrow(df_clean), " rows x ", ncol(df_clean), " cols\n")
cat("Data removed: ", round((1 - nrow(df_clean)/nrow(df)) * 100, 1), "%\n")

cat("\n========================================\n")

# Save cleaned data
# write_csv(df_clean, "data_cleaned.csv")

File: data-skill/templates/python/clean_data.py
#!/usr/bin/env python3
# Data Cleaning Template
# Purpose: Handle missing values, duplicates, and outliers

import pandas as pd
import numpy as np

# ============================================================================
# USER INPUT: Modify these variables
# ============================================================================
data_file = "data.csv"
# ============================================================================

# Load data
df = pd.read_csv(data_file)

print("\n" + "=" * 40)
print("DATA CLEANING REPORT")
print("=" * 40 + "\n")

# Initial dimensions
print(f"Original dimensions: {df.shape[0]} rows x {df.shape[1]} cols\n")

# ============================================================================
# 1. Remove Duplicates
# ============================================================================
df_clean = df.drop_duplicates()
removed_dup = len(df) - len(df_clean)
print(f"Duplicates removed: {removed_dup}")

# ============================================================================
# 2. Handle Missing Values
# ============================================================================
missing_by_col = df_clean.isna().sum()[df_clean.isna().sum() > 0].sort_values(ascending=False)

if len(missing_by_col) > 0:
    print("\nMissing values by column:")
    print(missing_by_col)
else:
    print("\nNo missing values found.")

# Option: Drop rows with missing values
# df_clean = df_clean.dropna()

# Option: Fill missing values
# for col in df_clean.select_dtypes(include=[np.number]).columns:
#     df_clean[col] = df_clean[col].fillna(df_clean[col].mean())

# ============================================================================
# 3. Detect Outliers (for numeric columns)
# ============================================================================
numeric_cols = df_clean.select_dtypes(include=[np.number]).columns.tolist()

if numeric_cols:
    print("\nOutlier detection (IQR method):")
    for col in numeric_cols:
        Q1 = df_clean[col].quantile(0.25)
        Q3 = df_clean[col].quantile(0.75)
        IQR = Q3 - Q1
        lower = Q1 - 1.5 * IQR
        upper = Q3 + 1.5 * IQR
        outliers = ((df_clean[col] < lower) | (df_clean[col] > upper)).sum()
        if outliers > 0:
            print(f"  {col}: {outliers} outliers")

# Final dimensions
print(f"\nClean dimensions: {df_clean.shape[0]} rows x {df_clean.shape[1]} cols")
print(f"Data removed: {(1 - len(df_clean)/len(df)) * 100:.1f}%")

print("\n" + "=" * 40)

# Save cleaned data
# df_clean.to_csv("data_cleaned.csv", index=False)
